ostd/src/boot/smp.rs

```diff
@@ -0,0 +1,146 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Symmetric multiprocessing (SMP) boot support.
+
+use alloc::collections::BTreeMap;
+use core::sync::atomic::{AtomicBool, Ordering};
+
+use spin::Once;
+
+use crate::{
+    arch::boot::smp::{bringup_all_aps, get_processor_info},
+    cpu,
+    mm::{
+        paddr_to_vaddr,
+        page::{self, meta::KernelMeta, ContPages},
+        PAGE_SIZE,
+    },
+    trap,
+};
+
+pub(crate) static AP_BOOT_INFO: Once<ApBootInfo> = Once::new();
+
+const AP_BOOT_STACK_SIZE: usize = PAGE_SIZE * 64;
+
+pub(crate) struct ApBootInfo {
+    /// It holds the boot stack top pointers used by all APs.
+    pub(crate) boot_stack_array: ContPages<KernelMeta>,
+    /// `per_ap_info` maps each AP's ID to its associated boot information.
+    per_ap_info: BTreeMap<u32, PerApInfo>,
+}
+
+struct PerApInfo {
+    is_started: AtomicBool,
+    // TODO: When the AP starts up and begins executing tasks, the boot stack will
+    // no longer be used, and the `ContPages` can be deallocated (this problem also
+    // exists in the boot processor, but the memory it occupies should be returned
+    // to the frame allocator).
+    boot_stack_pages: ContPages<KernelMeta>,
+}
+
+static AP_LATE_ENTRY: Once<fn() -> !> = Once::new();
+
+/// Boot all application processors.
+///
+/// This function should be called late in the system startup. The system must at
+/// least ensure that the scheduler, ACPI table, memory allocation, and IPI module
+/// have been initialized.
+///
+/// However, the function need to be called before any `cpu_local!` variables are
+/// accessed, including the APIC instance.
+pub fn boot_all_aps() {
+    // TODO: support boot protocols without ACPI tables, e.g., Multiboot
+    let Some(processor_info) = get_processor_info() else {
```

Comment:
`get_processor_info()` returns an `acpi::ProcessorInfo` which is unique to ACPI, or x86. It (involving a specific architecture) probably is not a good practice outside `arch` module.

Could we instead make a "owned" type like `PlatformInfo`, which provides an architecture-independent abstraction for different platforms? It could make it easier to support other boot protocols and architectures.